app/actions/ControlActions.js

```diff
@@ -924,3 +946,175 @@ export const getPeerInfo = () => (dispatch, getState) => {
+export const SETACCOUNTSPASSPHRASE_FAILED = "SETACCOUNTSPASSPHRASE_FAILED";
+export const SETACCOUNTSPASSPHRASE_SUCCESS = "SETACCOUNTSPASSPHRASE_SUCCESS";
+
+const setAccountPassphrase = async (
+  walletService,
+  accountNumber,
+  accountPassphrase,
+  newAcctPassphrase,
+  walletPassphrase
+) => {
+  try {
+    await wallet.setAccountPassphrase(
+      walletService,
+      accountNumber,
+      accountPassphrase,
+      newAcctPassphrase,
+      walletPassphrase
+    );
+    return;
+  } catch (error) {
+    return error;
+  }
+};
+
+export const setAccountsPass = (walletPassphrase) => async (
+  dispatch,
+  getState
+) => {
+  dispatch({ type: SETACCOUNTSPASSPHRASE_ATTEMPT });
+  try {
+    const oldAccounts = sel.balances(getState());
+    await wallet.unlockWallet(sel.walletService(getState()), walletPassphrase);
+    const accounts = Promise.all(
+      oldAccounts.map(async (acct) => {
+        // just skip if imported account.
+        if (acct.accountNumber === Math.pow(2, 31) - 1) {
+          return acct;
+        }
+        // we set the account passphrase as the wallet passphrase to avoid the user
+        // ending with multiple passphrases.
+
+        await setAccountPassphrase(
+          sel.walletService(getState()),
+          acct.accountNumber,
+          walletPassphrase,
+          walletPassphrase,
+          walletPassphrase
+        );
+        acct.encrypted = true;
+        return acct;
+      })
+    );
+    return dispatch({ type: SETACCOUNTSPASSPHRASE_SUCCESS, accounts });
+  } catch (error) {
+    dispatch({ error, type: SETACCOUNTSPASSPHRASE_FAILED });
+  }
+};
+
+export const UNLOCKACCTORWALLET_ATTEMPT = "UNLOCKACCTORWALLET_ATTEMPT";
+export const UNLOCKACCTORWALLET_FAILED = "UNLOCKACCTORWALLET_FAILED";
+export const UNLOCKACCTORWALLET_SUCCESS = "UNLOCKACCTORWALLET_SUCCESS";
+
+// unlockAcctAndExecFn unlocks the account and perform some acction. Locks the
+// account in case of success or error, if leaveUnlock is not informed.
+export const unlockAcctAndExecFn = (
+  passphrase,
+  acctNumber,
+  fn,
+  leaveUnlock
+) => async (dispatch, getState) => {
+  let res = null;
+  let fnError = null;
+  dispatch({ type: UNLOCKACCTORWALLET_ATTEMPT });
+
+  // sanity checks
+  const accounts = sel.balances(getState());
+
+  const account = accounts.find((acct) => acct.accountNumber === acctNumber);
+  if (!account) {
+    throw "Account not found";
+  }
+  if (!account.encrypted) {
+    throw "Account not encrypted";
+  }
+
+  // unlock wallet
+  try {
+    await wallet.unlockAccount(
+      sel.walletService(getState()),
+      passphrase,
+      acctNumber
+    );
+  } catch (error) {
+    // no need to lock as unlock errored.
+    dispatch({ type: UNLOCKACCTORWALLET_FAILED, error });
+    throw error;
+  }
+
+  // execute method
+  try {
+    res = await fn();
+  } catch (error) {
+    fnError = error;
+  }
+
+  if (fnError !== null) {
+    throw fnError;
+  }
```

Comment:
This will leave the account unlocked if `fnError` is filled in the block above. This needs to happen after the wallet is locked (if `leaveUnlock` is false).

Reply:
true, good catch.